scalyr_agent/platform_posix.py

```diff
@@ -634,10 +634,14 @@ def handle_terminate(signal_num, frame):
 
             It takes care of invoking the agent termination handler function.
             """
-            logger.debug("Received SIGTERM signal")
+            # Termination handler will cause log handlers to be closed so we need to check if they
+            # are still opened before trying to log a message.
+            if not scalyr_logging.HANDLERS_CLOSED:
```

Comment:
@Kami Are we getting an error if the handlers are closed?  I would think we want to handle that error down in the logger classes then.  If we are getting an error, it sounds like we are trying to write to a closed handler.

Reply:
Correct, we will see a bunch of messages like this on SIGTERM in some situations when agent is busy and doesn't shut down immediately in first SIGTERM (bb9035987994394fe9cd7104bbe39df26b5f8b53):

```bash
IOError: [Errno 0] Error
Logged from file platform_posix.py, line 637
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/logging/__init__.py", line 889, in emit
    stream.write(fs % msg)
```

When shutting down the agent using agent_main.py stop we try to send SIGTERM for up to 5 seconds (with 0.01 second delay in between) before sending SIGKILL if the agent still didn't shut down after this period has passed.

And the first SIGTERM signal invocation will cause log handlers to be closed so subsequent logger messages will fail.

Another option would also be to have some global variable which indicates if SIGTERM handler function has already be called and do a no-op if it has.

Reply:
But, if a log handler has been closed, why hasn't it been removed?  In my mental model of how the Python logging modules work... you insert messages to be logged, there's some dispatching logic, and a set of handlers decide which messages they will emit.  If we close a handler, I would think it should have been removed from the list of handlers that can emit logs.  Is that not the case, or am I not understanding something?